frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+data.
+
+Data Augmentation
+
+Besides loading, data augmentation expands datasets synthetically.
+Augmentations apply random transformations like flipping, cropping,
+rotating, altering color, adding noise etc. for images. For audio,
+common augmentations involve mixing clips with background noise, or
+modulating speed/pitch/volume.
+
+Augmentations increase variation in the training data. Frameworks like
+TensorFlow and PyTorch simplify applying random augmentations each epoch
+by integrating into the data pipeline.By programmatically increasing
+variation in the training data distribution, augmentations reduce
+overfitting and improve model generalization.
+
+Many frameworks make it easy to integrate augmentations into the data
+pipeline so they are applied on-the-fly each epoch. Together, performant
+data loaders and extensive augmentations enable practitioners to feed
+massive, varied datasets to neural networks efficiently. Hands-off data
+pipelines represent a significant improvement in usability and
+productivity. They allow developers to focus more on model architecture
+and less on data wrangling when training deep learning models.
+
+### Optimization Algorithms
+
+Training a neural network is fundamentally an iterative process that
+seeks to minimize a loss function. At its core, the goal is to fine-tune
+the model weights and parameters to produce predictions as close as
+possible to the true target labels. Machine learning frameworks have
+greatly streamlined this process by offering extensive support in three
+critical areas: loss functions, optimization algorithms, and
+regularization techniques.
+
+Loss Functions are useful to quantify the difference between the
+model\'s predictions and the true values. Different datasets require a
+different loss function to perform properly, as the loss function tells
+the computer the "objective" for it to aim to. Commonly used loss
+functions are Mean Squared Error (MSE) for regression tasks and
+Cross-Entropy Loss for classification tasks.
+
+Once the loss is computed, we need methods to adjust the model\'s
+parameters to reduce this loss or error during the training process. To
+do so, current frameworks use a gradient based approach, where it
+computes how much changes tuning the weights in a certain way changes
+the value of the loss function. Knowing this gradient, the model moves
+in the direction that reduces the gradient. There are many challenges
+associated with this, however, primarily stemming from the fact that the
+optimization problem is not convex, making it very easy to solve, and
+more details about this will come in the AI training sections. Modern
+frameworks come equipped with efficient implementations of several
+optimization algorithms, many of which are variants of gradient descent
+algorithms with stochastic methods and adaptive learning rates. More
+information can be found in the AI training section.
```

Comment:
It helps to illustrate the loss function and algorithms like gradient descent stochastic methods or learning rates with some equations. Again, I think this part is assuming too much prior knowledge, and those who are not familiar with the buzz words might not understand this paragraph.

Reply:
Good point. 

There will be an AI training chapter that will be coming in before this chapter; we are working on it (just not within the class context), and our hope is that that chapter would have set up the necessary background that would then flow into this chapter. but good feedback though. Thank you.